pre-docs/applying_manifest_blocks.md

```diff
@@ -0,0 +1,189 @@
+
+When the catalog compiles and is executed successfully on all targets, apply returns the reports generated by applying the catalog on each node.
+
+## Using Hiera Data
+
+Hiera is a tool for looking up data based on properties of a particular target. It allows you to separate configuration from context-specific data, such as facts or name of a target.
+
+Hiera data is used implicitly to override default class parameters. You can also explicitly lookup data from Hiera via lookup, for example
+```
+plan do_thing() {
+  apply('localhost') {
+    notice("Some data in Hiera: ${lookup('mydata')}")
+  }
+}
+```
+
+Manifest block compilation can access Hiera data added to your Bolt config. The default location for Hiera config is `$BOLTDIR/hiera.yaml`; this can be changed with the `hiera-config` key in a Bolt config file. Only Hiera version 5 is supported.
+
+Following the Hiera 5 convention, the default data dir is relative to `hiera.yaml` at `$BOLTDIR/data`. For config file examples, see https://puppet.com/docs/puppet/5.5/hiera_config_yaml_5.html.
+
+If a custom data provider is used (such as `hiera-eyaml`, which allows you to encrypt your data) the gem dependencies must be available to Bolt. If using `puppet-bolt` packages and installing `hiera-eyaml`
+- on Windows, run `"C:/Program Files/Puppet Labs/Bolt/bin/gem.bat" install hiera-eyaml`
+- on other platforms, run `/opt/puppetlabs/bolt/bin/gem install hiera-eyaml`
+
+## Available Plan Functions
+
+In addition to normal Puppet functions available to a catalog, such as `lookup`, some of Bolt's functions will also work in a manifest block
+- [puppetdb_query](plan_functions.md#puppetdb_query)
+- [puppetdb_facts](plan_functions.md#puppetdb_facts)
+- [get_targets](plan_functions.md#get_targets)
+- [facts](plan_functions.md#facts)
+- [vars](plan_functions.md#vars)
+
+## Options
+
+Apply supports common metaparameters available to Bolt functions:
+- `_catch_errors => true` will return a `ResultSet` including failed results, rather than failing the plan.
+- `_noop => true` will apply the manifest block in Puppet's noop mode, returning a report of the changes that would be made.
+- `_run_as => <user>` applies the manifest block as the specified user (on transports that support it).
+
+## Configuring Concurrency
+
+Each target requires a separate catalog be compiled with its unique facts and vars. Apply will compile and apply catalogs in parallel on the Bolt host. Concurrency of catalog compilation is controlled by a new `compile-concurrency` config option that's limited to 2x the number of threads your CPU can run concurrently, while catalog application uses Bolt's default thread pool controlled by the `concurrency` option.
+
+## Caveats
+
+Exported resources are not supported in apply blocks. Users should pass values directly instead of exporting and collecting them from PuppetDB.
```

Comment:
Should we mention the puppetdb query function here or does that just confuse things?

If you need to interact with resources managed during a normal puppet run you can use puppetdb_query

Reply:
I'm not sure that's a common scenario, but I can include it.